diff -rupN tomcat-6.0.35-src.org/java/org/apache/catalina/connector/Connector.java tomcat-6.0.35-src/java/org/apache/catalina/connector/Connector.java
--- tomcat-6.0.35-src.org/java/org/apache/catalina/connector/Connector.java	2011-11-28 02:22:45.000000000 -0800
+++ tomcat-6.0.35-src/java/org/apache/catalina/connector/Connector.java	2013-06-27 16:50:06.157278174 -0700
@@ -126,6 +126,11 @@ public class Connector
      */
     protected boolean xpoweredBy = false;
 
+    /*
+    * Is trim the space between the ; and charset in content type header
+    */
+    protected boolean trimContentType = true;
+
 
     /**
      * Descriptive information about this Connector implementation.
@@ -895,6 +900,14 @@ public class Connector
         setProperty("xpoweredBy", String.valueOf(xpoweredBy));
     }
 
+    public boolean isTrimContentType() {
+        return trimContentType;
+    }
+
+    public void setTrimContentType(boolean trimContentType) {
+        this.trimContentType = trimContentType;
+    }
+
     /**
      * Enable the use of IP-based virtual hosting.
      *
diff -rupN tomcat-6.0.35-src.org/java/org/apache/catalina/connector/CoyoteAdapter.java tomcat-6.0.35-src/java/org/apache/catalina/connector/CoyoteAdapter.java
--- tomcat-6.0.35-src.org/java/org/apache/catalina/connector/CoyoteAdapter.java	2011-11-28 02:22:45.000000000 -0800
+++ tomcat-6.0.35-src/java/org/apache/catalina/connector/CoyoteAdapter.java	2013-06-27 17:33:07.689159077 -0700
@@ -252,6 +252,8 @@ public class CoyoteAdapter implements Ad
     	                org.apache.coyote.Response res)
         throws Exception {
 
+        res.setConnector(connector);
+
         Request request = (Request) req.getNote(ADAPTER_NOTES);
         Response response = (Response) res.getNote(ADAPTER_NOTES);
 
diff -rupN tomcat-6.0.35-src.org/java/org/apache/coyote/Response.java tomcat-6.0.35-src/java/org/apache/coyote/Response.java
--- tomcat-6.0.35-src.org/java/org/apache/coyote/Response.java	2011-11-28 02:22:45.000000000 -0800
+++ tomcat-6.0.35-src/java/org/apache/coyote/Response.java	2013-06-27 17:34:56.725154047 -0700
@@ -20,6 +20,7 @@ package org.apache.coyote;
 import java.io.IOException;
 import java.util.Locale;
 
+import org.apache.catalina.connector.Connector;
 import org.apache.tomcat.util.buf.ByteChunk;
 import org.apache.tomcat.util.http.MimeHeaders;
 
@@ -50,7 +51,6 @@ public final class Response {
      */
     private static Locale DEFAULT_LOCALE = Locale.getDefault();
 
-
     // ----------------------------------------------------- Instance Variables
 
     /**
@@ -89,6 +89,9 @@ public final class Response {
     protected boolean commited = false;
 
 
+    protected Connector connector;
+
+
     /**
      * Action hook.
      */
@@ -226,8 +229,15 @@ public final class Response {
         this.commited = v;
     }
 
+    public Connector getConnector() {
+        return connector;
+    }
 
-    // -----------------Error State --------------------
+    public void setConnector(Connector connector) {
+        this.connector = connector;
+    }
+
+// -----------------Error State --------------------
 
 
     /** 
@@ -520,10 +530,14 @@ public final class Response {
 
         String ret = contentType;
 
-        if (ret != null 
+        if (ret != null
             && characterEncoding != null
             && charsetSet) {
-            ret = ret + ";charset=" + characterEncoding;
+            if (connector != null && !connector.isTrimContentType()) {
+                ret = ret + "; charset=" + characterEncoding;
+            } else {
+                ret = ret + ";charset=" + characterEncoding;
+            }
         }
 
         return ret;
