diff -ur apache-tomcat-6.0.24-src.orig/java/org/apache/coyote/http11/Http11AprProcessor.java apache-tomcat-6.0.24-src/java/org/apache/coyote/http11/Http11AprProcessor.java
--- apache-tomcat-6.0.24-src.orig/java/org/apache/coyote/http11/Http11AprProcessor.java	2010-01-19 05:43:40.000000000 -0800
+++ apache-tomcat-6.0.24-src/java/org/apache/coyote/http11/Http11AprProcessor.java	2010-06-09 17:12:33.000000000 -0700
@@ -1754,13 +1754,13 @@
                 (inputFilters[Constants.CHUNKED_FILTER]);
             contentDelimitation = true;
         } else {
-            for (int i = 2; i < inputFilters.length; i++) {
+            /*for (int i = 2; i < inputFilters.length; i++) {
                 if (inputFilters[i].getEncodingName()
                     .toString().equals(encodingName)) {
                     inputBuffer.addActiveFilter(inputFilters[i]);
                     return true;
                 }
-            }
+            }*/
             return false;
         }
         return true;
diff -ur apache-tomcat-6.0.24-src.orig/java/org/apache/coyote/http11/Http11NioProcessor.java apache-tomcat-6.0.24-src/java/org/apache/coyote/http11/Http11NioProcessor.java
--- apache-tomcat-6.0.24-src.orig/java/org/apache/coyote/http11/Http11NioProcessor.java	2010-01-19 05:43:40.000000000 -0800
+++ apache-tomcat-6.0.24-src/java/org/apache/coyote/http11/Http11NioProcessor.java	2010-06-09 17:12:44.000000000 -0700
@@ -1781,13 +1781,13 @@
                 (inputFilters[Constants.CHUNKED_FILTER]);
             contentDelimitation = true;
         } else {
-            for (int i = 2; i < inputFilters.length; i++) {
+            /*for (int i = 2; i < inputFilters.length; i++) {
                 if (inputFilters[i].getEncodingName()
                     .toString().equals(encodingName)) {
                     inputBuffer.addActiveFilter(inputFilters[i]);
                     return true;
                 }
-            }
+            }*/
             return false;
         }
         return true;
diff -ur apache-tomcat-6.0.24-src.orig/java/org/apache/coyote/http11/Http11Processor.java apache-tomcat-6.0.24-src/java/org/apache/coyote/http11/Http11Processor.java
--- apache-tomcat-6.0.24-src.orig/java/org/apache/coyote/http11/Http11Processor.java	2010-01-19 05:43:40.000000000 -0800
+++ apache-tomcat-6.0.24-src/java/org/apache/coyote/http11/Http11Processor.java	2010-06-09 17:04:38.000000000 -0700
@@ -1660,13 +1660,13 @@
                 (inputFilters[Constants.CHUNKED_FILTER]);
             contentDelimitation = true;
         } else {
-            for (int i = 2; i < inputFilters.length; i++) {
+            /*for (int i = 2; i < inputFilters.length; i++) {
                 if (inputFilters[i].getEncodingName()
                     .toString().equals(encodingName)) {
                     inputBuffer.addActiveFilter(inputFilters[i]);
                     return true;
                 }
-            }
+            }*/
             return false;
         }
         return true;
Only in apache-tomcat-6.0.24-src: output
