Index: java/org/apache/catalina/connector/Connector.java
===================================================================
--- java/org/apache/catalina/connector/Connector.java	(revision 1541661)
+++ java/org/apache/catalina/connector/Connector.java	(working copy)
@@ -126,7 +126,12 @@
      */
     protected boolean xpoweredBy = false;
 
+    /*
+    * Is trim the space between the ; and charset in content type header
+    */
+    protected boolean trimContentType = true;
 
+
     /**
      * Descriptive information about this Connector implementation.
      */
@@ -895,6 +900,14 @@
         setProperty("xpoweredBy", String.valueOf(xpoweredBy));
     }
 
+    public boolean isTrimContentType() {
+        return trimContentType;
+    }
+
+    public void setTrimContentType(boolean trimContentType) {
+        this.trimContentType = trimContentType;
+    }
+
     /**
      * Enable the use of IP-based virtual hosting.
      *
Index: java/org/apache/catalina/connector/CoyoteAdapter.java
===================================================================
--- java/org/apache/catalina/connector/CoyoteAdapter.java	(revision 1541661)
+++ java/org/apache/catalina/connector/CoyoteAdapter.java	(working copy)
@@ -252,6 +252,8 @@
     	                org.apache.coyote.Response res)
         throws Exception {
 
+        res.setConnector(connector);
+
         Request request = (Request) req.getNote(ADAPTER_NOTES);
         Response response = (Response) res.getNote(ADAPTER_NOTES);
 
Index: java/org/apache/catalina/core/StandardContextValve.java
===================================================================
--- java/org/apache/catalina/core/StandardContextValve.java	(revision 1541661)
+++ java/org/apache/catalina/core/StandardContextValve.java	(working copy)
@@ -114,16 +114,6 @@
     public final void invoke(Request request, Response response)
         throws IOException, ServletException {
 
-        // Disallow any direct access to resources under WEB-INF or META-INF
-        MessageBytes requestPathMB = request.getRequestPathMB();
-        if ((requestPathMB.startsWithIgnoreCase("/META-INF/", 0))
-            || (requestPathMB.equalsIgnoreCase("/META-INF"))
-            || (requestPathMB.startsWithIgnoreCase("/WEB-INF/", 0))
-            || (requestPathMB.equalsIgnoreCase("/WEB-INF"))) {
-            notFound(response);
-            return;
-        }
-
         // Wait if we are reloading
         boolean reloaded = false;
         while (context.getPaused()) {
Index: java/org/apache/coyote/Response.java
===================================================================
--- java/org/apache/coyote/Response.java	(revision 1541661)
+++ java/org/apache/coyote/Response.java	(working copy)
@@ -20,6 +20,7 @@
 import java.io.IOException;
 import java.util.Locale;
 
+import org.apache.catalina.connector.Connector;
 import org.apache.tomcat.util.buf.ByteChunk;
 import org.apache.tomcat.util.http.MimeHeaders;
 
@@ -89,6 +90,9 @@
     protected boolean commited = false;
 
 
+    protected Connector connector;
+
+
     /**
      * Action hook.
      */
@@ -226,7 +230,14 @@
         this.commited = v;
     }
 
+    public Connector getConnector() {
+        return connector;
+    }
 
+    public void setConnector(Connector connector) {
+        this.connector = connector;
+    }
+
     // -----------------Error State --------------------
 
 
@@ -523,7 +534,11 @@
         if (ret != null 
             && characterEncoding != null
             && charsetSet) {
-            ret = ret + ";charset=" + characterEncoding;
+            if (connector != null && !connector.isTrimContentType()) {
+            ret = ret + "; charset=" + characterEncoding;
+            } else {
+                ret = ret + ";charset=" + characterEncoding;
+            }
         }
 
         return ret;
