Index: java/org/apache/coyote/Response.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- java/org/apache/coyote/Response.java	(revision 1616579)
+++ java/org/apache/coyote/Response.java	(revision )
@@ -21,6 +21,7 @@
 import java.io.StringReader;
 import java.util.Locale;
 
+import org.apache.catalina.connector.Connector;
 import org.apache.tomcat.util.buf.ByteChunk;
 import org.apache.tomcat.util.http.MimeHeaders;
 import org.apache.tomcat.util.http.parser.HttpParser;
@@ -92,6 +93,9 @@
     protected boolean commited = false;
 
 
+    protected Connector connector;
+
+
     /**
      * Action hook.
      */
@@ -229,7 +233,14 @@
         this.commited = v;
     }
 
+    public Connector getConnector() {
+        return connector;
+    }
 
+    public void setConnector(Connector connector) {
+        this.connector = connector;
+    }
+
     // -----------------Error State --------------------
 
 
@@ -496,7 +507,11 @@
         if (ret != null
             && characterEncoding != null
             && charsetSet) {
+            if (connector != null && !connector.isTrimContentType()) {
-            ret = ret + ";charset=" + characterEncoding;
+            ret = ret + "; charset=" + characterEncoding;
+            } else {
+                ret = ret + ";charset=" + characterEncoding;
+            }
         }
 
         return ret;
Index: java/org/apache/catalina/connector/CoyoteAdapter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- java/org/apache/catalina/connector/CoyoteAdapter.java	(revision 1616579)
+++ java/org/apache/catalina/connector/CoyoteAdapter.java	(revision )
@@ -252,6 +252,8 @@
     	                org.apache.coyote.Response res)
         throws Exception {
 
+        res.setConnector(connector);
+
         Request request = (Request) req.getNote(ADAPTER_NOTES);
         Response response = (Response) res.getNote(ADAPTER_NOTES);
 
Index: java/org/apache/catalina/connector/Response.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- java/org/apache/catalina/connector/Response.java	(revision 1616579)
+++ java/org/apache/catalina/connector/Response.java	(revision )
@@ -1017,7 +1017,7 @@
      * @param cookie Cookie to be added
      */
     public void addCookieInternal(final Cookie cookie) {
-        addCookieInternal(cookie, false);
+        addCookieInternal(cookie, cookie.isHttpOnly());
     }
 
     /**
Index: java/org/apache/catalina/connector/Connector.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- java/org/apache/catalina/connector/Connector.java	(revision 1616579)
+++ java/org/apache/catalina/connector/Connector.java	(revision )
@@ -129,7 +129,12 @@
      */
     protected boolean xpoweredBy = false;
 
+    /*
+    * Is trim the space between the ; and charset in content type header
+    */
+    protected boolean trimContentType = true;
 
+
     /**
      * Descriptive information about this Connector implementation.
      */
@@ -964,6 +969,14 @@
     public void setXpoweredBy(boolean xpoweredBy) {
         this.xpoweredBy = xpoweredBy;
         setProperty("xpoweredBy", String.valueOf(xpoweredBy));
+    }
+
+    public boolean isTrimContentType() {
+        return trimContentType;
+    }
+
+    public void setTrimContentType(boolean trimContentType) {
+        this.trimContentType = trimContentType;
     }
 
     /**
Index: java/org/apache/catalina/core/StandardContextValve.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- java/org/apache/catalina/core/StandardContextValve.java	(revision 1616579)
+++ java/org/apache/catalina/core/StandardContextValve.java	(revision )
@@ -114,16 +114,6 @@
     public final void invoke(Request request, Response response)
         throws IOException, ServletException {
 
-        // Disallow any direct access to resources under WEB-INF or META-INF
-        MessageBytes requestPathMB = request.getRequestPathMB();
-        if ((requestPathMB.startsWithIgnoreCase("/META-INF/", 0))
-            || (requestPathMB.equalsIgnoreCase("/META-INF"))
-            || (requestPathMB.startsWithIgnoreCase("/WEB-INF/", 0))
-            || (requestPathMB.equalsIgnoreCase("/WEB-INF"))) {
-            notFound(response);
-            return;
-        }
-
         // Wait if we are reloading
         boolean reloaded = false;
         while (context.getPaused()) {
Index: java/javax/servlet/http/Cookie.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- java/javax/servlet/http/Cookie.java	(revision 1616579)
+++ java/javax/servlet/http/Cookie.java	(revision )
@@ -84,11 +84,12 @@
     private String comment;	// ;Comment=VALUE ... describes cookie's use
 				// ;Discard ... implied by maxAge < 0
     private String domain;	// ;Domain=VALUE ... domain that sees cookie
+    private boolean httpOnly;
     private int maxAge = -1;	// ;Max-Age=VALUE ... cookies auto-expire
     private String path;	// ;Path=VALUE ... URLs that see the cookie
     private boolean secure;	// ;Secure ... e.g. use SSL
     private int version = 0;	// ;Version=1 ... means RFC 2109++ style
-    
+
     
 
     /**
@@ -231,8 +232,13 @@
 	return domain;
     }
 
+    public boolean isHttpOnly() {
+        return httpOnly;
+    }
 
-
+    public void setHttpOnly(final boolean httpOnly) {
+        this.httpOnly = httpOnly;
+    }
 
     /**
      * Sets the maximum age of the cookie in seconds.
